docs/protocol.md

```diff
@@ -0,0 +1,286 @@
+  | ConfirmAck | 5     |
+
+The Extensions field currently only contains the 8-bit block type, in the highest
+8 bits of the field. But this is only the case if the message contains a Block (i.e. Publish, ConfirmReq and ConfirmAck messages).
+See `rai::message::block_type_set()` in `rai/node/common.cpp`.
+
+## Keepalive message
+
+Reference: class `rai::keepalive` in `rai/node/common.cpp`
+
+A Keepalive message contains a list of IP addresses of peers, in the form of (IPv6 address, port) pairs.
+Each IP address is stored as 16 bytes, each port as 2 bytes:
+
+  | Field        | Size (bytes) | Description     |
+  | ------------ | ------------ | --------------- |
+  | IPv6 address | 16           | Big-endian?     |
+  | Port         | 2            | Little-endian?! |
+
+The number of pairs in the message can vary, from 1 up to 8 peers.
+
+## Publish message
+
+Reference: class `rai::publish` in `rai/node/common.cpp`
+
+A Publish message simply contains a published Block, which was either created or
+forwarded by another node.
+
+See below for the way a block in encoded.
+
+## ConfirmReq message
+
+Reference: class `rai::confirm_req` in `rai/node/common.cpp`
+
+XXX A ConfirmReq message contains a Block for which a vote has been requested?
+
+  | Field | Size (bytes) | Description        |
+  | ----- | ------------ | ------------------ |
+  | Block | Varies       | A serialized Block |
+
+## ConfirmAck message
+
+Reference: class `rai::confirm_ack` in `rai/node/common.cpp`
+
+A ConfirmAck message contains a Vote by a representative on a Block.
+
+  | Field | Size (bytes) | Description       |
+  | ----- | ------------ | ----------------- |    
+  | Vote  | (varies)     | A serialized Vote |
+
+
+## Vote
+
+Reference: class `rai::vote` in `rai/common.cpp`
+
+  | Field     | Size (bytes) | Description                          |
+  | --------- | ------------ | ------------------------------------ |
+  | Account   | 32           | Public key of the voting Account     |
+  | Signature | 64           | Signature                            |
+  | Sequence  | 8            | 64-bit sequence number (big-endian?) |
+  | Block     | Varies       | The Block being voted on             |
+
+The Signature of a Vote signs the following list of hashables: `hash(Block)`, `Sequence`
+
+XXX what is the exact function of the sequence number?
+
+# Blocks
+
+Reference: `rai/lib/blocks.cpp`
+
+Blocks are encoded into a byte stream, with each field directly following
+the previous value. All blocks contain a Signature and a Work value, which follow
+the block-specific fields:
+
+  | Field     | Size (bytes) | Description                                |
+  | --------- | ------------ | ------------------------------------------ |
+  | *Block-specific fields* | |                                           |
+  | Signature | 64           | Ed25519 signature of the block's hashables |
+  | Work      | 8            | Unsigned 64-bit integer (little endian?)   |
```

Comment:
Work for universal blocks (state_blocks) are big endian.

Reply:
Why is the work value stored as uint64_t in the node anyway? It seem to be only ever used a input for the blake2b hash and is never used as integer value.

Reply:
@paulmelis Nodes use the value to validate the work. As long as it's a hexadecimal value with a length of 16 characters, storing it as 64bit binary makes the most sense to me since it's the most compact one i think. What better options do we have?

Reply:
Length 8 byte array (`uint8_t work[8]`)

Reply:
Casting a uint64_t to a uint8_t* as is currently done makes it dependent on the endianness of the system running on (although there aren’t many BE cpus around). 

Also, reading/writing the value as an array of bytes is easier, it saves the current endian handling (which was changed for state blocks)